File: typescript/src/tracing.ts
/**
 * OpenTelemetry tracing setup for the beval framework.
 *
 * Provides tracing configuration for process graders that inspect execution
 * traces. See SPEC.md §9 (Process Graders via Traces) and spec/otel-conventions.md.
 */

import { trace, type Tracer } from "@opentelemetry/api";
import { NodeTracerProvider } from "@opentelemetry/sdk-trace-node";
import {
  InMemorySpanExporter,
  SimpleSpanProcessor,
} from "@opentelemetry/sdk-trace-node";

/**
 * Configure OpenTelemetry tracing with an in-memory exporter.
 *
 * Returns the exporter so process graders can inspect captured spans.
 */
export function setupTracing(serviceName = "beval"): InMemorySpanExporter {
  const exporter = new InMemorySpanExporter();
  const provider = new NodeTracerProvider();
  provider.addSpanProcessor(new SimpleSpanProcessor(exporter));
  provider.register();
  return exporter;
}

/** Get a tracer instance from the current provider. */
export function getTracer(name = "beval"): Tracer {
  return trace.getTracer(name);
}

File: go/tracing.go
package beval

// Tracer holds tracing state for the beval framework.
//
// Will use: go.opentelemetry.io/otel and go.opentelemetry.io/otel/sdk
type Tracer struct {
	ServiceName string
}

// SetupTracing configures OpenTelemetry tracing with an in-memory exporter.
//
// Returns a Tracer that process graders can use to inspect captured spans.
// See SPEC §9 (Process Graders via Traces).
//
// Will use: go.opentelemetry.io/otel/sdk/trace with InMemoryExporter
func SetupTracing(serviceName string) *Tracer {
	// Stub: returns a Tracer with the service name.
	return &Tracer{ServiceName: serviceName}
}

// GetTracer returns a tracer instance for the given name.
//
// Will use: go.opentelemetry.io/otel.GetTracerProvider().Tracer(name)
func GetTracer(name string) *Tracer {
	// Stub: returns a new Tracer.
	return &Tracer{ServiceName: name}
}
